eventlet/hubs/epolls.py

```diff
@@ -29,3 +31,11 @@ def add(self, evtype, fileno, cb, tb, mac):
 
     def do_poll(self, seconds):
         return self.poll.poll(seconds)
+
+    def remove(self, listener):
+        if all([len(x) == 0 for x in self.listeners.values()]):
+            return
+        super().remove(listener)
+        if all([len(x) == 0 for x in self.listeners.values()]):
+            self.poll.close()
```

Comment:
Imagine workflow:

- talk some network
- local processing
- at some point there is no more listeners so epoll gets closed
- talk some network again ...

Imagine this goes in hot loop of single use create/close, like DNS client or kind of monitoring agent.

Long story short, last fd closed is not a good signal for resource cleanup.

End of thread should call `Hub.abort()` which is proper way to release all hub resources.

AFAIK, Python memory management should call `Hub.__del__` when it goes out of reach of non-weak references. So fd leak may be due to dangling reference or we need `__del__` of `self.abort()`

Reply:
ok, so i have now a working version. The important part in the test is the call to `hubs._threadlocal.hub.abort(wait=True)`.
Without this call (or without the `wait=True`) we do not switch to the `BaseHub.run` method and allow it to exit.
When the thread then exits the `BaseHub.run` method still references `self` (so the Hub) causing it to never be garbage collected.

However i'm still not happy with this solution as it requires an explicit call to `abort(wait=True)` at the end of your thread. This is not easily possible if e.g. eventlet is used in a logging library and therefor called without the main code knowing about eventlet at all

Reply:
actually i just realized that we do not leak file descriptors at all if `abort(wait=True)` is called. It seems like the default epoll implementation already closes the filedescriptor for us

Reply:
i would therefor close this PR. Thanks very much for all the information you provided

Reply:
> However i'm still not happy with this solution as it requires an explicit call to abort(wait=True) at the end of your thread.

Me too! If you want to spend more time, these things should help:

- `sys.getrefcount(hub)` returns internal reference counter. It's +1 to call getrefcount function, so 1 is the lowest possible number for dead object and 2 means something else holds ref to object.
- `gc.get_referrers` basically is what you need, but read description, it will not show some refs.
- `objgraph` library.

Reply:
Thanks for the hints. While i will not work on it right away i'll keep it in my mind if we hit this issue again